src/async/mod.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO: remove following line once all changes are merged into master
+#![allow(dead_code)]
+
+mod promise;
+mod lock;
+
+use std::fmt::{self, Debug, Formatter};
+use std::sync::Arc;
+
+use futures::{Async, Future, Poll};
+use futures::task::{self, Task};
+
+use error::{Error, Result};
+use self::lock::SpinLock;
+use self::promise::Shutdown as ShutdownPromise;
+
+
+/// A handle that is used to notify future that the task finishes.
+pub struct NotifyHandle<T> {
+    result: Option<Result<T>>,
+    park: Option<Task>,
+    stale: bool,
+}
+
+impl<T> NotifyHandle<T> {
+    fn new() -> NotifyHandle<T> {
+        NotifyHandle {
+            result: None,
+            park: None,
+            stale: false,
+        }
+    }
+
+    /// Set the result and notify future if necessary.
+    fn result(&mut self, res: Result<T>) {
+        self.result = Some(res);
+
+        if let Some(ref t) = self.park {
+            t.unpark();
+        }
+    }
+}
+
+type Inner<T> = SpinLock<NotifyHandle<T>>;
+
+fn new_inner<T>() -> Arc<Inner<T>> {
+    Arc::new(SpinLock::new(NotifyHandle::new()))
+}
+
+/// A future object for task that is scheduled to `CompletionQueue`.
+pub struct CqFuture<T> {
+    inner: Arc<Inner<T>>,
+}
+
+impl<T> CqFuture<T> {
+    fn new(inner: Arc<Inner<T>>) -> CqFuture<T> {
+        CqFuture { inner: inner }
+    }
+}
+
+impl<T> Future for CqFuture<T> {
+    type Item = T;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<T, Error> {
+        let mut guard = self.inner.lock();
+        if guard.stale {
+            return Err(Error::FutureStale);
```

Comment:
Instead of returning an error `FutureStale`, why not panic at line 84. Since this future has already resolved, polling it again does not make sense.